add information to dev-mode and astarte_dev_tool (#1067)
astarte-platform/astarte: doc/pages/development-guide/001-dev_guide.md

This guide is focused on writing and testing changes to Astarte. It assumes basic knowledge in the following areas:

-   You have read the [contributing guidelines](https://github.com/astarte-platform/astarte/blob/master/CONTRIBUTING.md) for the project
-   You are familiar with basic Astarte concepts such as Realms, Interfaces, Devices and Triggers. If you are not, you can learn more on the [documentation](001-intro_architecture.md)!
-   You have already run at least once [Astarte in 5 minutes](010-astarte_in_5_minutes.md)
-   If you plan to work on the code, you have some knowledge of Elixir and Docker

### Platforms

> Currently, the development of [Astarte project](https://www.github.com/astarte-platform/astarte) is tested for the following platforms:

Reviewer: ```suggestion
Currently, the development of [Astarte project](https://www.github.com/astarte-platform/astarte) is tested on the following platforms:
```

-   Ubuntu linux >= 20.04, amd64/arm64
-   Debian linux >= bullseye, amd64/arm64

> -   macOS amd64/arm64, via Microsoft VSCode devcontainer (preferred)

Reviewer: As far as I know, we tested _only_ with devcontainers on macOS
```suggestion
-   macOS amd64/arm64, via Microsoft VSCode devcontainer
```

Reviewer 2: devcontainers are available also on different editors than VSCode (e.g. https://github.com/nohzafk/emacs-devcontainer). I would keep the notion that VSCode is the preferred way of running devcontainers, but stress that this is only for macOS. The `macOS amd64/arm64, via Microsoft VSCode devcontainer (preferred)` line might be interpreted as "devcontainers on vscode on macOs are the preferred way to run Astarte" which is not true. Matbe something like `(preferred for arm development)`?

Author: @Annopaolo no worries, "preferred" will be removed

Finally, Astarte is backed by three libraries:

-   [Astarte Core](https://github.com/astarte-platform/astarte_core): contains definitions common to the application, e.g. Realm name, Device ID validity, Interface definitions etc.
-   [Astarte RPC](https://github.com/astarte-platform/astarte_rpc): contains all the Protobufs and code used in the internal RPC mechanism. In most cases, RPCs are between an API component and the related backend.

> -   [Astarte Data Access](https://github.com/astarte-platform/astarte_data_access): contains some interactions with the database that are common to all Astarte services, e.g. queries for interface values.

Reviewer: also here

Reviewer 2: See comment above

The main Astarte repo contains a number of tools that can be used everyday when interacting with Astarte. They are found in the `tools` directory subdirectory of the [main Astarte repository](https://www.github.com/astarte-platform/astarte) on Github:

-   Astarte Device Fleet Simulator: create a fleet of virtual devices. Mainly used for load testing
-   Astarte E2E: a tool that sends data from a virtual device and checks the resulting value in Astarte using triggers. Mainly used for monitoring
-   Astarte Import and Astarte Export: used to load/export data into/from Astarte

> -   Astarte Dev Tool: development helper, facilitating the up & running of the platform in development mode (see [astarte_dev_tool](#astarte_dev_tool))

Reviewer: ```suggestion
-   Astarte Dev Tool: Spin and set up Astarte in development mode with ease (see [astarte_dev_tool](#astarte_dev_tool))
```

## Environment

There are two possible alternatives for development, depending on whether you want to use a complete setup on the development host or prefer to use devcontainers.

> The choice also falls on whether you want to [work on a single component](#testing-a-single-component) or prefer to pull up the entire Astarte platform.

Reviewer: the phrasal  verb "to pull up " generally means to stop near something (https://www.wordreference.com/enit/pull%20up), in this context something like "to put up the entire Astarte Platform" is probably more appropriate if we do not want to repeat setup (preferrable imho)

> The choice also falls on whether you want to [work on a single component](#testing-a-single-component) or prefer to pull up the entire Astarte platform.

Reviewer: the phrasal verb "to pull up " generally means to stop near something or to physically raise something (https://www.wordreference.com/enit/pull%20up), in this context something like "to put up/spin up/ start up the entire Astarte Platform" is probably more appropriate if we do not want to repeat setup (preferrable imho)

### Devcontainers (preferred on macOs, requires Microsoft VSCode as IDE)

The development system only requires _docker_ and _docker compose_ to be installed. macOS users normally prefer to use _Docker Desktop_, in which case _docker compose_ is automatically installed.
The development containers will already have the necessary add-ons to optimise the work.

#### Important Notice

> Development via devcontainer is _only possible_ with regard to the [apps](https://www.github.com/astarte-platform/astarte) of Astarte, and once started, all other containers of the platform will be started automatically.
> If you wish to develop on the tools or other libraries (e.g. [Astarte Core](https://github.com/astarte-platform/astarte_core)), you must proceed by the next alternative.

Reviewer: Content is ok, imho can be rephrased to be more coincise and direct
```suggestion
When using dev containers, it is only possible to work on Astarte  [apps](https://www.github.com/astarte-platform/astarte), with the main requirement that all the Astarte containers must be started.
If you want to work on the other tools or libraries (e.g. [Astarte Core](https://github.com/astarte-platform/astarte_core)), you must use another set-up method.
```

### Host with Elixir installed

In order to install and manage Elixir (and Erlang), we recommend using either the [asdf](https://asdf-vm.com/) runtime version manager or the [nix](https://nixos.org/) package manager.

#### Using asdf

From inside the main Astarte directory, install the Elixir and Erlang asdf plugins

```bash
asdf plugin add erlang
asdf plugin add elixir
```

Then install Elang/OTP and Elixir with

```bash
asdf install
```

#### Using nix

From inside the main Astarte directory, just run

-   `nix develop`, if you are using flakes
-   `nix-shell`, if not

## Development on the entire platform

> The easiest way to develop on the whole platform, loading everything Astarte locally, is to start the various containers of _docker compose_ in "development" mode. In this configuration, you can then start the `watch` of the changes, to allow hot-code reloading.

Reviewer: ```suggestion
The easiest way to develop on the whole platform is to start the project (and, thus, every astarte container) with _docker compose_ in "development" mode.
This way you can also enable the hot reloading mode with the 'watch' command of [astarte_dev_tool](https://www.github.com/astarte-platform/astarte) or directly with _docker compose_.
```

One can take advantage of [astarte_dev_tool](https://www.github.com/astarte-platform/astarte) or act directly on _docker compose_.

### <a name="astarte_dev_tool"></a>astarte_dev_tool

> From inside the [astarte_dev_tool](https://www.github.com/astarte-platform/astarte) directory, `mix` to build the tool itself, followed by the commands to use it, in which the root of the Astarte project must be specified with the _path_ argument:

Author: @eddbbt See row 72

> From inside the main Astarte directory, install the Elixir and Erlang asdf plugins

> From inside the [astarte_dev_tool](https://www.github.com/astarte-platform/astarte) directory, `mix` to build the tool itself, followed by the commands to use it, in which the root of the Astarte project must be specified with the _path_ argument:

Reviewer: @shinnokdisengir those are two completely different contexts

> From inside the [astarte_dev_tool](https://www.github.com/astarte-platform/astarte) directory, `mix` to build the tool itself, followed by the commands to use it, in which the root of the Astarte project must be specified with the _path_ argument:

Reviewer 2: It is better to be explicit on commands, even tough `mix` is just an alias for `mix build` (TIL!). What about
```suggestion
From inside the astarte_dev_tool directory, `mix build` to build the tool itself, followed by the commands to use it. Note that the root Astarte directory must be explicitly specified with `-p`: 
```

(also, I suggest defaulting the `-p` flag to `../..` in the dev tool, so that most of the time it can be omitted)

> From inside the [astarte_dev_tool](https://www.github.com/astarte-platform/astarte) directory, `mix` to build the tool itself, followed by the commands to use it, in which the root of the Astarte project must be specified with the _path_ argument:

Reviewer 2: Also, let's be consistent here and make sure everything code-related is denoted with the right syntax (either single or triple `backticks`)

-   `mix astarte_dev_tool.system.up -p <path>`, starts the platform and all containers in "dev-mode".
-   `mix astarte_dev_tool.system.watch -p <path>`, starts the "watcher", which allows hot-code reloading in case of code changes. The process is stopped and must be explicitly closed with `ENTER`.
-   `mix astarte_dev_tool.system.down -p <path> [-v]`, to terminate the platform. The `-v` parameter will also delete the volumes of the respective containers.

### docker compose

From inside the Astarte directory:

-   `docker compose -f docker-compose.yml -f docker-compose.dev.yml up --build [-d]`, starts the platform and all containers in "dev-mode".
-   `docker compose -f docker-compose.dev.yml -f docker-compose.dev.yml watch --no-up`, starts the watching. The process must be terminated explicitly with `Ctrl+C`.
-   `docker compose -f docker-compose.dev.yml -f docker-compose.dev.yml down [-v]`, to terminate the platform. The `-v` parameter will also delete the volumes of the respective containers.

## Development on individual components

> It is also possible not to start the entire platform, which is done automatically both with the procedures listed above and with the use of _devcontainers_.

Reviewer: maybe it's better  to describe how to do, not how to not do, so:
```suggestion
It is also possible not to start the entire platform, to do so enter the directory of the service you are interested in and run the appropriate mix command (varies from service to service).
Usually you don't want to do so, though 
```

- Astarte Device Fleet Simulator: create a fleet of virtual devices. Mainly used for load testing
- Astarte E2E: a tool that sends data from a virtual device and checks the resulting value in Astarte using triggers. Mainly used for monitoring
- Astarte Import and Astarte Export: used to load/export data into/from Astarte
- Astarte Dev Tool: development helper, facilitating the up & running of the platform in development mode (see [astarte_dev_tool](#astarte_dev_tool))

## Environment

There are two possible alternatives for development, depending on whether you want to use a complete setup on the development host or prefer to use devcontainers.
The choice also falls on whether you want to [work on a single component](#testing-a-single-component) or prefer to pull up the entire Astarte platform.

### Devcontainers (preferred on macOs, requires Microsoft VSCode as IDE)

The development system only requires _docker_ and _docker compose_ to be installed. macOS users normally prefer to use _Docker Desktop_, in which case _docker compose_ is automatically installed.
The development containers will already have the necessary add-ons to optimise the work.

> #### Important Notice

Reviewer 2: Let's be consistent: use a `[!NOTE]`  tag as done below

If you wish to develop on the tools or other libraries (e.g. [Astarte Core](https://github.com/astarte-platform/astarte_core)), you must proceed by the next alternative.

### Host with Elixir installed

In order to install and manage Elixir (and Erlang), we recommend using either the [asdf](https://asdf-vm.com/) runtime version manager or the [nix](https://nixos.org/) package manager.

#### Using asdf

From inside the main Astarte directory, install the Elixir and Erlang asdf plugins

```bash
asdf plugin add erlang
asdf plugin add elixir
```

Then install Elang/OTP and Elixir with

```bash
asdf install
```

#### Using nix

From inside the main Astarte directory, just run

- `nix develop`, if you are using flakes
- `nix-shell`, if not

## Development on the entire platform

The easiest way to develop on the whole platform, loading everything Astarte locally, is to start the various containers of _docker compose_ in "development" mode. In this configuration, you can then start the `watch` of the changes, to allow hot-code reloading.
One can take advantage of [astarte_dev_tool](https://www.github.com/astarte-platform/astarte) or act directly on _docker compose_.

> ### <a name="astarte_dev_tool"></a>astarte_dev_tool

Reviewer 2: `<a>`s are automatically generated by ex_docs, no need to put them manually

> [!NOTE]
> Some services are a bit special (for now!) and might need a little more setup for testing.
>
> -   AppEngine API needs the AMQP exchange `astarte_events` to be declared:
>     after having started RabbitMQ, run the following lines to declare it:
>
> ```bash
>  docker exec $RABBITMQ_CONTAINER_NAME rabbitmqadmin declare exchange name=astarte_events type=direct
> ```

> -   Pairing needs a CFSSL instance available and exposed on port 8080: you can run
>
> ```bash
> docker run --net=host -p 8080/tcp ispirata/docker-alpine-cfssl-autotest:astarte
> ```

> >

Reviewer: I can't see any meaningful change here, pls avoid unnecessary noise

-   Create a JWT wth claims on AppEngine, Pairing and RealmManagement APIs:
    `astartectl utils gen-jwt all-realm-apis -k $REALM_PRIVATE_KEY_FILE`
-   Start a virtual device:
    `docker run --net="host" -e "DEVICE_ID=$(astartectl utils device-id generate-random)" -e "PAIRING_URL=http://api.astarte.localhost/pairing" -e "REALM=$REALM_NAME" -e "PAIRING_JWT=$(astartectl utils gen-jwt pairing -k $REALM_PRIVATE_KEY_FILE)" -e "IGNORE_SSL_ERRORS=true" astarte/astarte-stream-qt5-test:1.0.4` - In this case, two interfaces need to be already installed in your realm: `org.astarte-platform.genericsensors.Values` and `org.astarte-platform.genericcommands.ServerCommands` - You can find them in the `standard-interfaces` directory in the Astarte repo

### Sending data from Astarte to a device

Using `astartectl` you can send messages to a device. For example:

-   Publish a (server-owned) datastream:
    `astartectl appengine devices publish-datastream $DEVICE_ID $SERVER_OWNED_DATASTREAM_INTERFACE $PATH $VALUE -r $REALM_NAME -k $REALM_PRIVATE_KEY_FILE -u "http://api.astarte.localhost/"`
-   Set a (server-owned) property:

>     `astartectl appengine devices set-property $DEVICE_ID $SERVER_OWNED_PROPERTY_INTERFACE $PATH $VALUE -r $REALM_NAME -k $REALM_PRIVATE_KEY_FILE -u "http://api.astarte.localhost/"`

Reviewer: pls avoid noise

Reviewer 2: It is a good idea to conform to a standard markdown formatting. However, in order to have self-contained commits, let's separate changes in the file formatting from the ones to the actual content. I suggest splitting the commit so

Reviewer 2: After IRL discussion, we decided to standardize doc formatting in another PR, and bring in only content changes in this one

Now, some commands you might find helpful to set up a realm:

- Create a realm keypair:

>     `astartectl utils gen-keypair $REALM_NAME`

Reviewer 2: Consistency: let's not add more whitespaces, as agreed
